### PR TITLE
Master

### PR DESCRIPTION
**Pull request by novaksam**
I’ve expanded the functionality of the script a bit to enable more granular policy and extension attribute creation, that way we don’t have to create templates for each piece of software the comes along and needs them, we can just modify the recipe.

### CODE REVIEW

File: example_templates/PolicyTemplate.xml [hunk 1]
```diff
@@ -2,27 +2,33 @@
...
         <category>
             <name>%POLICY_CATEGORY%</name>
         </category>
     </general>
-    <scope>
-        <!--Scope added by JSSImporter-->
-    </scope>
-    <package_configuration>
-        <!--Package added by JSSImporter-->
+	<scope>
```
Comment (novaksam) [inline example_templates/PolicyTemplate.xml]: Not sure what happened here; I'm guessing it autoformatted the file or something.

**Comment by sheagcraig** [comment]
@novaksam I appreciate all of the work you've done on this-it looks pretty comprehensive and well thought out.

That being said, it goes against the original design of JSSImporter that I worked out with the AutoPkg guys. Our goal in creating the template system was to avoid creating input variables to control every aspect of the workflow. And in practice, I think this decision was correct in that I don't *ever* have to edit my policy template to accommodate new software.

As you have picked up on, this isn't necessarily the case with the group, script, and extension attribute templates. That being said, I find that I only have to modify the group, script, and extension attribute templates when I'm working on a recipe that needs an extension attribute or script, which is only two or three recipes at this point, in my repo. On the other hand, for people with more heavily customized workflows (who use an EA or script with *every* policy), I can see where this would be helpful.

It seems like there's no harm in adding optional input variables to the replacement_dict for text substitution for those who want to utilize it. The text replacement and the templates can't conflict with each other, which is nice.

On the other hand, it adds a ton more processor arguments. Also, in "properly" written recipes, these arguments should be duplicated in the input variable section so they can be overridden. I'm not sure that the ability to collect all of the configuration in a single file is worth the increased confusion of that many arguments, when most of that information is more concisely reflected in the XML templates.

Also, when people share their recipes, it may mean a lot of overriding to use someone's recipe which is heavily configured for their environment, whereas now it would be simple to just replace the policy template with your environment's standard template.

Finally, I'm hesitant to change the names of the existing extension attribute keys, forcing all existing recipes to update the names. Also, scripts, extension attributes, and groups allow an arbitrary number number of entries. However, if you specify the name and template path in the input variables and then use those values for ALL members of the array, you break that ability (i.e. each extension attribute in the array will have the same name and use the same template path, which is not what you would want.) This could still work, but you would need to retain the existing processor arguments in ADDITION to the ones you've added, making things even more involved.

So I need some time to think about this.

One possible solution is to think through a way to do this in a "automagical" way to begin with. You may or may not be aware (I just kind of figured it out myself the other day) that AutoPkg will add whatever input variables you want into the env of the processors. Meaning, for example, I can specify an argument key for a processor named ```EXAMPLE_KEY_THAT_DOESNT_EXIST``` and set the value to ```Tacos``` or whatever, and when the recipe runs, sure enough, that key and value are added to the env.

If JSSImporter just added ALL key/value pairs from the env to the replacement dictionary and replaced things whenever it found something, that could allow you to specify whatever you wanted without changing the way JSSImporter currently works.

The potential risks/downsides of this are that *who knows* what *other* processors/recipes/overrides are adding to the env? What if there are conflicting values from the ones specified with the JSSImporter?

And again, to make your recipes more valuable to the community, they need to have the processor arguments overrideable by the input variables, which results in a lot of duplication in values again.

**Comment by novaksam** [comment]
You bring up some really good points; I think it's just in my nature to have as much flexibility and customization as possible, but I can see how some of my proposals are scoop creepy.

Most of my inspiration for doing this was out of frustration with existing JSS recipes, which would usually have their own templates for each recipe which prevented simply over-riding the recipe (I use AutoPkgr) from being, in my opinion, as streamlined as it could be.

I was thinking creating more flexible templates that could be modified by the recipe itself would make it simpler to build new recipes and easier to modify them for each environment, but thinking about it some more (after reading your message above) I think there is a way to strip down some of the complexity I added in while still being able to use templates for extension and group creation.

I'll modify the code a little bit to explain what I'm thinking.

**Comment by sheagcraig** [comment]
Cool deal. Looking forward to seeing what you come up with. Having other people hacking on this is both exciting and scary! I have a responsibility to all of these people using this whom I've never met!

I'm interested by what you said about each recipe in AutoPkgr having its own templates? My intention was always for their to be ONE template for all policies, and ONE for groups. Additional ones should only be needed for "problem" recipes. Am I missing something in how AutoPkgr downloads or overrides recipes that is creating all of this redundancy?

**Comment by novaksam** [comment]
Well when you create an override with AutoPkgr it doesn't copy over templates or images into the override directory, so I have to go and copy them in myself (though honestly it's probably just an issue in their code).

An example of recipes that each have their own templates is in the golbiga-recipes repo in the JSS folder. For the AdobeAIR jss recipe there is a recipe, a picture, an extension attribute template, and a smart group template, just for that one piece of software; I believe that this is needed with the current version of JSSImporter because of the use of an extension attribute.

**Comment by novaksam** [comment]
Looks like there might need to be a little work done on updating extension attributes:

Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1469, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1463, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 1287, in run_recipes
    autopackager.process(recipe)
  File "/Library/AutoPkg/autopkglib/__init__.py", line 466, in process
    self.env = processor.process()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 295, in process
    self.main()
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 697, in main
    self.extattrs = self.handle_extension_attributes()
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 553, in handle_extension_attributes
    added_env="jss_extension_attribute_updated")
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 516, in _update_or_create_new
    recipe_object.save()
  File "/Library/Python/2.7/site-packages/python_jss-0.5.5-py2.7.egg/jss/jss.py", line 626, in save
    "Object Conflict! If trying to post a new "
jss.exceptions.JSSPostError: Object Conflict! If trying to post a new object, look for name conflict and delete.

I'll try to look into this next week.

**Comment by novaksam** [comment]
I think I've found the source of the issue I was getting.

I looks like (at line 486) when the call is made to self.j.factory.get_object(obj_cls, name) it's not taking the variables I set in the templates into account (and why would it, it's not suppose to work that way)

One possible solution is to write in an if statement at that call and call it differently if obj_cls is jss.computerextensionattribute.

Another possible solution is to modify the call to _update_or_create_new in handle_extension_attributes.

A very interesting predicament...

**Comment by novaksam** [comment]
Or possibly reading the value of <name></name> out of the extension attribute template, and swapping the variables that are passed in handle_extension_attributes

**Comment by novaksam** [comment]
The commit I just pushed fixed the issue.

**Comment by sheagcraig** [comment]
@novaksam There's a funny thing in Extension Attributes that doesn't really make sense. There's a ```name``` input variable, and then the name is specified in the extension attribute template. It's completely possible to have them specified as different names. When you try to post to name1 and the template says the name is name2, there's a 409 Conflict error.

I'm working on the next release, which drops the ```name``` input variable and just uses the template.

**Comment by sheagcraig** [comment]
@novaksam Take a look at the latest testing version.

I made a pretty big change to the replacement dictionary building.

The replacement_dict is a dictionary that is used to replace text in the templates. It looks for, for example, %VERSION% and replaces it with the value of the version. Before, it was built from a handful of input variables to JSSImporter.

Now, it copies every string/unicode item from the AutoPkg env into the replacement dictionary, and then for compatibility purposes, does the "official" JSSImporter input variables.

What this means is that you can specify an input variable pretty much anywhere in the recipe (well, specifically in the "Input" section OR the arguments to JSSImporter) and it will get added to the replacement dictionary.

Example:
If I add to my jss recipe JSSImporter arguments the lines:
```
<key>NUM_OF_TRIFORCE</key>
<string>3</string>
```
and then in, for example, the PolicyTemplate.xml, you use the text replacement tag in a value:
```
<policy>
    <general>
        <name>Install Latest %PROD_NAME%</name>
        <enabled>true</enabled>
        <frequency>Ongoing</frequency>
        <category>
            <name>%POLICY_CATEGORY%</name>
        </category>
    </general>
    <scope>
        <!--Scope added by JSSImporter-->
    </scope>
    <package_configuration>
        <!--Package added by JSSImporter-->
    </package_configuration>
    <scripts>
        <!--Scripts added by JSSImporter-->
    </scripts>
    <self_service>
        <use_for_self_service>true</use_for_self_service>
        <install_button_text>Install %NUM_OF_TRIFORCE%</install_button_text>
        <self_service_description>%SELF_SERVICE_DESCRIPTION%</self_service_description>
    </self_service>
    <maintenance>
        <recon>true</recon>
    </maintenance>
</policy>
```
it should merrily swap the values in.

This should allow you to arbitrarily add in whatever variables you need, rather than having to add in an expanded vocabulary.

Which is not to say that we *shouldn't* expand the vocabulary, but it seems like a better way to handle the text replacement overall. It may still make sense to ALSO formalize some more values.

**Comment by novaksam** [comment]
I think that this will work for me :) 

I'm also thinking if there's a way to mark packages to "Install on reboot" so that scripts run correctly if the machine were being imaged, but that's a whole different thing.

I think you can close this pull request if you'd like.

**Comment by sheagcraig** [comment]
@novaksam Just a little update-
The newest version also handles finding the icons, templates, etc, following a specified search-order through first the *calling*-recipe's folder, then parent folder, then parent recipe's folder and parent folder, etc. This means you don't have to specify relative paths to files in your recipes (just give the filename), and you'll no longer have to copy those files into your override directory if you're not changing those files.

**Pull request closed without merging**